tools/Az.Tools.Predictor/Az.Tools.Predictor/AzPredictor.cs

```diff
@@ -56,16 +56,19 @@ public sealed class AzPredictor : ICommandPredictor
 
         private readonly IAzPredictorService _service;
         private readonly ITelemetryClient _telemetryClient;
+        private readonly Settings _settings;
 
         /// <summary>
         /// Constructs a new instance of <see cref="AzPredictor"/>
         /// </summary>
         /// <param name="service">The service that provides the suggestion</param>
         /// <param name="telemetryClient">The client to collect telemetry</param>
-        public AzPredictor(IAzPredictorService service, ITelemetryClient telemetryClient)
+        /// <param name="settings">The settings of the service</param>
+        public AzPredictor(IAzPredictorService service, ITelemetryClient telemetryClient, Settings settings)
         {
             this._service = service;
             this._telemetryClient = telemetryClient;
+            this._settings = settings;
         }
```

Comment:
The `StartEarlyProcessing` seems to be doing a lot redundant work. The list of history passed in is a moving window, so every time this method gets called, the latest command line is added to the list and the oldest command line is removed, so ideally, you only need to parse the new one, which should be at the index 0 of the list.

Reply:
I agree with what you said that we have redundant work there. I'll try to reduce that part of work.